[FLINK-24932] Support build FRocksDB Java on Apple silicon

Myasuka opened the pull request:
This PR selected 4 necessary commits to enable building FRocksDB Java on Apple silicon.

jay-zhuang committed:
Fix clang13 build error (#9374)

Summary: Pull Request resolved: https://github.com/facebook/rocksdb/pull/9374

Test Plan: Add CI for clang13 build

Reviewed By: riversand963

Differential Revision: D33522867

Pulled By: jay-zhuang

fbshipit-source-id: 642756825cf0b51e35861fb847ebaee4611b76ca

jay-zhuang committed:
Update circleci xcode version (#9405)

Summary:
xcode 11.3.1 is deprecated https://circleci.com/docs/2.0/testing-ios/ , jobs are failing:
```
failed to create host: Image xcode:11.3.0 is not supported
```

Pull Request resolved: https://github.com/facebook/rocksdb/pull/9405

Test Plan: CI

Reviewed By: ajkr, hx235

Differential Revision: D33674462

Pulled By: jay-zhuang

fbshipit-source-id: 85dd27aad84d26eaaa5c5375015344182b2c50b9

adamretter committed:
Fix RocksJava releases for macOS (#9662)

Summary:
Addresses the problems described in https://github.com/facebook/rocksdb/pull/9254#issuecomment-1054598516 and https://github.com/facebook/rocksdb/pull/9254#issuecomment-1059574837 that have blocked a RocksJava release

**NOTE** Also needs to be ported to 6.29.fb branch.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/9662

Reviewed By: ajkr

Differential Revision: D34689200

Pulled By: pdillinger

fbshipit-source-id: c62fe34c54f05be5a00ee1daec8ec7454baa5eb8

Myasuka committed:
Add support for Apple Silicon to RocksJava (#9254)

    Summary:
    Fixes facebook/rocksdb#7720

    Updated Makefile with flags to define target architecture when compiling/linking,
    and added goal `rocksdbjavastaticosxub` to build a OS X Universal Binary native library.

    Pull Request resolved: https://github.com/facebook/rocksdb/pull/9254

    Reviewed By: mrambacher

    Differential Revision: D33551160

    Pulled By: pdillinger

    fbshipit-source-id: 9ce9962e03aacf55014545a6cdf638b5b14b8fa9

Myasuka committed:
Resolve the broken download link of zlib-1.2.11

Myasuka committed:
[hotfix] Resolve brken make format

Myasuka commented:
@carp84 would you please take a look at this PR?
Since all necessary work to support building on apple silicon has been supported on c++ side in RocksDB-6.20.3, we only need to cherry-pick java side work to ensure it could build binaries on macos. And since we only need changes on java side, we do not need to update binaries on other platforms.

Myasuka commented:
Apart from asking resources within Flink community, @adamretter , who is responsible for RocksJava in RocksDB community, would you please help to review this PR? 
This PR only includes picked changes from the RocksDB community to support running Flink forked RocksDB-6.20.3 on apple silicon chips.

adamretter commented:
@Myasuka I'm not really sure what you would like me to contribute? I'm mostly just a little responsible for Java and multi-arch in RocksDB. Can you tell me how this is related?

Myasuka commented:
@adamretter Thanks for your response. 
This repo is used as the Flink community forked RocksDB. We want to support RocksDB on apple silicon chips in the coming flink-1.16 release. However, Flink community lacks resources to help review RocksDB-related code currently, and since commits in this PR mainly reviewed or created by you, I think you're certainly cable of reviewing this.

The reasons why we did not bump to RocksDB-6.29.5:
1. We found a bit of performance regression in our Flink benchmark compared with RocksDB-6.20.3 with RocksDB-6.29.5.
2. We just want to avoid generating too many binaries on different platforms (we always suffer from this, I'm not sure whether you could share experiences here), and keep the code at RocksDB-6.20.3 and only cherry-pick changes on building and java side could help this. Thus, we only need to publish the jar package with updated java classes and binary `.so` files on mac-os platform.

Hope you could take a look at this PR, thanks!

adamretter commented:
@Myasuka thank you, it's very kind of you to ask. However, I am afraid that I don't have enough time to also contribute to Flink.

Myasuka commented:
@adamretter Thanks for your response. The feature freeze data of flink-1.16 is 2022-08-09, and you can still feel free to review this PR before that time.

This PR consists of three commits from the RocksDB community:
1. https://github.com/facebook/rocksdb/pull/9254 
2. https://github.com/facebook/rocksdb/pull/9662
3. https://github.com/facebook/rocksdb/pull/9374
The other two minor commits are introduced to keep zlib version as 1.2.11 and resolve the broken code format.

fredia commented:
Hi @Myasuka, I opened circleci, can you submit this PR again to trigger tests?

Myasuka committed:
Upgrade to Ubuntu 20.04 in our CircleCI config

Summary:
Ubuntu 16.04 has reached EOL. The patch upgrades the image for all of
our CircleCI jobs to the latest, namely `ubuntu-2004:202111-02`.

Myasuka committed:
Disable useless broken tests due to ci-image upgraded

Since we do not need to publish new binaries on linux/windows platforms,
we can disable the broken tests due to ubuntu imaged upgraded in circle.

rkhachatryan commented on .circleci/config.yml:
```diff
@@ -504,13 +538,38 @@ jobs:
+          name: "Set Java Environment"
+          command: |
+            echo "JAVA_HOME=${JAVA_HOME}"
+            echo 'export PATH=$JAVA_HOME/bin:$PATH' >> $BASH_ENV
```
NIT: double-quote `$BASH_ENV`?

rkhachatryan commented on .circleci/config.yml:
```diff
@@ -596,12 +655,14 @@ workflows:
-  build-linux-encrypted-env:
-    jobs:
-      - build-linux-encrypted-env
+# Temporarily disable CI tests on mem-env as we would not update linux binaries in FLINK-24932.
```
Do you think it makes sense to create a ticket to re-enable it later (and reference the ticket here)?
ditto: `build-linux-gcc-4_8-no_test_run`, `build-cmake-mingw`

Myasuka replied:
I think this could be done once we make FLINK-24932 merged with the current solution.

rkhachatryan commented on Makefile:
```diff
@@ -301,6 +305,19 @@ ifeq ($(LIB_MODE),shared)
+# linking with jemalloc (as it won't be arm64-compatible) and remove some other options
+# set during platform detection
+DISABLE_JEMALLOC=1
+PLATFORM_CFLAGS := $(filter-out -march=native -DHAVE_SSE42 -DHAVE_AVX2, $(PLATFORM_CFLAGS))
+PLATFORM_CXXFLAGS := $(filter-out -march=native -DHAVE_SSE42 -DHAVE_AVX2, $(PLATFORM_CXXFLAGS))
```
I'm wondering whether `filter-out` can remove too much here. Shouldn't it be replaced with `filter`?

rkhachatryan commented on Makefile:
```diff
@@ -1968,7 +1987,7 @@ SHA256_CMD = sha256sum
 ZLIB_VER ?= 1.2.11
 ZLIB_SHA256 ?= c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1
-ZLIB_DOWNLOAD_BASE ?= http://zlib.net
+ZLIB_DOWNLOAD_BASE ?= https://web.archive.org/web/20211202213624/http://www.zlib.net
```
Shouldn't it be `https://zlib.net/fossils/`?

Myasuka replied:
Actually, using `web.archive.org` to download the legacy version is referred from RocksDB community. Thanks for the information on this link!

rkhachatryan reviewed (approved):
Thanks a lot for the PR @Myasuka ,
I have only minor comments, otherwise LGTM given that the build succeeds.

fredia commented:
@Myasuka , because we use the [free plan of circleCI](https://circleci.com/pricing/), we can only use S⁠–⁠L size resource size, could you please update the `.circleci/config.yaml`, and due to the limit of [memory](https://app.circleci.com/pipelines/github/ververica/frocksdb/57/workflows/2f1ab659-ff66-4d02-9ee4-f610fdaa54e5/jobs/1873) and [space](https://app.circleci.com/pipelines/github/ververica/frocksdb/57/workflows/3edf0acc-3364-4201-b7ee-fdaa742c1aae/jobs/1865), some of tests can't be finished, could we ignore these tests?

Myasuka committed:
[hotfix] Use zlib's fossils page to replace web.archive

fredia commented on .circleci/config.yml:
```diff
@@ -120,7 +139,9 @@ executors:
 jobs:
   build-macos:
```
`build-macos` failed because of `no output time`, could you please [increase timeout](https://support.circleci.com/hc/en-us/articles/360007188574-Build-has-Hit-Timeout-Limit#:~:text=CircleCI%20has%20a%20built-in,will%20be%20canceled%20and%20stopped.) here:
```
- run: 
    no_output_timeout: 30m
```

fredia commented:
@Myasuka Thanks for the update, the `build-macos` failed because of `no output time`, could you please [increase timeout](https://support.circleci.com/hc/en-us/articles/360007188574-Build-has-Hit-Timeout-Limit#:~:text=CircleCI%20has%20a%20built-in,will%20be%20canceled%20and%20stopped.) at `build-macos` job:
```
- run: 
    no_output_timeout: 30m
```

Myasuka committed:
[hotfix] Change the resource request when running CI
